File: sdk/src/main/java/com/careagle/sdk/base/fragment/BaseRecyclerViewFragment.java
package com.careagle.sdk.base.fragment;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.careagle.sdk.R;
import com.careagle.sdk.base.IBaseRecyclerView;
import com.careagle.sdk.weight.EmptyView;
import com.chad.library.adapter.base.BaseQuickAdapter;

import java.util.ArrayList;
import java.util.List;


/**
 * Created by lida on 2018/1/15.
 */

public abstract class BaseRecyclerViewFragment<T> extends BaseCompatFragment implements IBaseRecyclerView<T>, SwipeRefreshLayout.OnRefreshListener, EmptyView.ReloadListener, BaseQuickAdapter.RequestLoadMoreListener {

    private int initialPage = 1;
    protected int page = initialPage;
    protected int pageSize = 10;
    protected EmptyView emptyView;
    protected ArrayList<T> list;

    public abstract RecyclerView getPageRecyclerView();

    public abstract BaseQuickAdapter getAdapter();

    public abstract SwipeRefreshLayout getPageSwipeRefreshLayout();

    public EmptyView getEmptyView() {
        return emptyView;
    }

    public List<T> getList() {
        return list;
    }

    public abstract void initData();

    public abstract void initLogic();

    @Override
    public void initUI(View view, @Nullable Bundle savedInstanceState) {

    }

    @Override
    public int getPage() {
        return page;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initRecyclerView();
    }

    public void initRecyclerView() {
        list = new ArrayList<>();
        initLogic();
        getPageSwipeRefreshLayout().setColorSchemeColors(getResources().getColor(R.color.theme_color));
        getPageSwipeRefreshLayout().setOnRefreshListener(this);
        getPageRecyclerView().setLayoutManager(getLinearLayoutManager() == null ? new LinearLayoutManager(getActivity()) : getLinearLayoutManager());
        getPageRecyclerView().setAdapter(getAdapter());
        emptyView = new EmptyView(getActivity());
        getAdapter().setEmptyView(getEmptyView());
        getAdapter().setHeaderAndEmpty(isEmptyAndHeader());
        if (isCanLoadMore()) {
            getAdapter().setOnLoadMoreListener(this, getPageRecyclerView());
        }
        getEmptyView().setLoadState(EmptyView.LoadState.LOAD_STATE_LOADING);
        getEmptyView().setReloadListener(this);
    }

    public LinearLayoutManager getLinearLayoutManager() {
        return null;
    }

    public boolean isEmptyAndHeader() {
        return false;
    }

    /**
     * 下拉刷新
     */
    @Override
    public void onRefresh() {
        page = initialPage;
        initData();
    }

    /**
     * 是否可以上啦加载更多
     *
     * @return
     */
    public boolean isCanLoadMore() {
        return true;
    }

    /**
     * 加载数据
     *
     * @param totalPage 总页数
     * @param data      数据
     */
    public void setData(int totalPage, List<T> data) {
        if (page == initialPage) getList().clear();
        if (data != null) {
            getList().addAll(data);
        }
        //page < totalPage &&
        if (isCanLoadMore() && (data != null && data.size() == pageSize)) {
            getAdapter().loadMoreComplete();
        } else {
            getAdapter().loadMoreEnd();
        }
        if (getPageSwipeRefreshLayout() != null)
            getPageSwipeRefreshLayout().setRefreshing(false);
        if (list.size() == 0) {
            getEmptyView().setLoadState(EmptyView.LoadState.LOAD_STATE_EMPTY);
        }
        getAdapter().notifyDataSetChanged();
    }

    /**
     * 加载失败  如果第一页就失败  那就显示用布局的错误
     */
    @Override
    public void showLoadError() {
        getAdapter().loadMoreFail();
        if (getPageSwipeRefreshLayout() != null) {
            getPageSwipeRefreshLayout().setRefreshing(false);
        }
        if (page > initialPage) {
            page--;
        }
        if (getList().size() == 0) {
            getEmptyView().setLoadState(EmptyView.LoadState.LOAD_STATE_ERROR);
        }
    }

    /**
     * 上拉加载更多
     */
    @Override
    public void onLoadMoreRequested() {
        page++;
        initData();
    }

    /**
     * 点击空布局加载失败
     */
    @Override
    public void onReload() {
        onRefresh();
    }
}

File: app/src/main/java/com/xfb/xinfubao/fragment/BaseRecyclerViewFragment.kt
package com.xfb.xinfubao.fragment

import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import com.careagle.sdk.R
import com.careagle.sdk.weight.EmptyView
import com.chad.library.adapter.base.BaseQuickAdapter
import com.chad.library.adapter.base.BaseViewHolder
import java.util.*

/**
 * 列表fragment
 */
abstract class BaseRecyclerViewFragment<T> : BaseFragment(),
    SwipeRefreshLayout.OnRefreshListener, BaseQuickAdapter.RequestLoadMoreListener,
    EmptyView.ReloadListener {

    //初始页数
    protected val initialPage = 1
    protected var page = initialPage
    protected var pageSize = 10
    protected var emptyView: EmptyView? = null
    protected var list = ArrayList<T>()

    abstract fun pageRecyclerView(): RecyclerView?

    abstract fun pageAdapter(): BaseQuickAdapter<T, BaseViewHolder>?

    abstract fun pageSwipeRefreshLayout(): SwipeRefreshLayout?

    abstract fun initData()

    abstract fun initLogic()

    open fun pageEmptyView(): EmptyView? {
        if (emptyView == null && context != null) {
            emptyView = EmptyView(context!!)
        }
        return emptyView
    }

    fun pageList(): ArrayList<T> {
        return list
    }

    override fun onLazyLoad() {
        initRecyclerView()
    }

    private fun initRecyclerView() {
        initLogic()
        emptyView = pageEmptyView()
        pageSwipeRefreshLayout()?.setColorSchemeColors(resources.getColor(R.color.theme_color))
        pageSwipeRefreshLayout()?.setOnRefreshListener(this)
        pageRecyclerView()?.layoutManager = pageLayoutManager()
        pageRecyclerView()?.adapter = pageAdapter()
        pageAdapter()?.emptyView = emptyView
        if (isCanLoadMore()) {
            pageAdapter()?.setOnLoadMoreListener(this, pageRecyclerView())
        }
        emptyView?.setLoadState(EmptyView.LoadState.LOAD_STATE_LOADING)
        emptyView?.setReloadListener(this)
    }

    open fun pageLayoutManager(): RecyclerView.LayoutManager {
        return LinearLayoutManager(activity)
    }

    open fun loadData(data: List<T>, isEndPage: Boolean? = null) {
        if (page == initialPage) list.clear()
        list.addAll(data)
        var canLoadMore = isCanLoadMore() && (isEndPage ?: (data.size >= pageSize))
        if (isEndPage != null) {
            canLoadMore = !isEndPage
        }
        if (canLoadMore) {
            pageAdapter()?.loadMoreComplete()
        } else {
            pageAdapter()?.loadMoreEnd()
        }
        pageSwipeRefreshLayout()?.isRefreshing = false
        if (list.size == 0) {
            emptyView?.setLoadState(EmptyView.LoadState.LOAD_STATE_EMPTY)
        }
        pageAdapter()?.notifyDataSetChanged()
        if (page == 0) {
            pageRecyclerView()?.scrollToPosition(0)
        }
    }

    /**
     * 是否可以上啦加载更多
     *
     * @return
     */
    open fun isCanLoadMore(): Boolean {
        return true
    }

    /**
     * 加载失败  如果第一页就失败  那就显示用布局的错误
     */
    fun showLoadError() {
        pageAdapter()?.loadMoreFail()
        pageSwipeRefreshLayout()?.isRefreshing = false
        if (page > 0) {
            page--
        }
        if (list.size == 0) {
            emptyView?.setLoadState(EmptyView.LoadState.LOAD_STATE_ERROR)
        }
    }

    override fun onRefresh() {
        page = initialPage
        initData()
    }

    override fun onLoadMoreRequested() {
        page++
        initData()
    }

    override fun onReload() {
        onRefresh()
    }

    fun notifyDataSetChanged() {
        pageAdapter()?.notifyDataSetChanged()
        if (list.size == 0) {
            emptyView?.setLoadState(EmptyView.LoadState.LOAD_STATE_EMPTY)
        }
    }
}